src/server/scene/surface_stack.cpp

```diff
@@ -493,6 +493,40 @@ void ms::SurfaceStack::send_to_back(const mir::scene::SurfaceSet &ss)
     }
 }
 
+auto ms::SurfaceStack::is_above(std::weak_ptr<scene::Surface> const& a, std::weak_ptr<scene::Surface> const& b) const -> bool
+{
+    RecursiveReadLock lg(guard);
+    if (a.expired())
+        return false;
+    else if (b.expired())
+        return true;
+
+    auto const shared_a = a.lock();
+    auto const shared_b = b.lock();
+    if (shared_a->depth_layer() < shared_b->depth_layer())
+        return false;
+    else if (shared_a->depth_layer() > shared_b->depth_layer())
+        return true;
+
+    auto const& layer = surface_layers[shared_a->depth_layer()];
+    std::optional<size_t> a_index;
+    std::optional<size_t> b_index;
+    for (size_t surface_index = 0; surface_index < layer.size(); surface_index++)
+    {
+        if (shared_a == layer[surface_index])
+            a_index = surface_index;
+        else if (shared_b == layer[surface_index])
+            b_index = surface_index;
+    }
+
+    if (!a_index.has_value())
+        return false;
+    else if (!b_index.has_value())
+        return true;
+
+    return a_index.value() > b_index.value();
```

Comment:
I don't see that iterating over the whole layer every time is significantly better than calling `std::find()` twice.

In any case, isn't this faster?

```suggestion
    bool found_b = false;
    for (auto const& entry : layer)
    {
        if (entry == b) found_b = true;
        if (entry == a) return found_b;
    }
    
    return false;
```